File: jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        if (arr.size() == 0) 
            return false;
    
        return jump(arr, start);
    }
    
    bool jump(vector<int>& arr, int index) {
        if (index < 0 || index >= arr.size() || arr[index] < 0) //check if already visited
            return false;
        
        if (arr[index] == 0)
            return true;
        
        //mark as visited:
        arr[index] *= -1;
        
        return jump(arr, index-arr[index]) || jump(arr, index+arr[index]);    
    }
};

File: jump-game-iii/jump-game-iii.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        return canReach(arr, start, new HashSet());
    }
    
    public boolean canReach(int[] arr, int idx, Set<Integer> set) {
        if (idx < 0 || idx >= arr.length || set.contains(idx))
            return false;
        
        if (arr[idx] == 0)
            return true;
        
        set.add(idx);
        return canReach(arr, idx + arr[idx], set) || canReach(arr, idx - arr[idx], set);
    }
}